### ISSUE TITLE
wai request logger middleware does not catch all exceptions

### ISSUE THREAD

**Issue by mwotton** [yesodweb/wai#496]
```
import           Network.Wai.Handler.Warp
import           Network.Wai.Middleware.RequestLogger

main :: IO ()
main = run 3000 $  logStdoutDev $ appLogsOk


appLogsOk _ respond = respond (throw $ userError "baby did a bad bad thing")
appNoLog _ respond = throw $ userError "baby did a bad bad thing"
```

The first one logs the incoming request, the second one doesn't. (Obviously this is terrible code, but it's derived from a problem I had in a servant webapp, and not seeing the incoming request in the logs made it a lot harder to debug.)

**Comment by snoyberg** [comment]
This seems like reasonable behavior, I'm not sure what real world use case
would expect something different

**Comment by mwotton** [comment]
I disagree - no matter how bad the code behind it is, the logger should always show the received request.

**Comment by snoyberg** [comment]
I'm open to a pull request that fixes this, assuming it doesn't impact
performance

**Comment by mwotton** [comment]
ok, i'll have a go.

**Comment by fizruk** [comment]
I have just bumped into this when an SQL exception didn't result in a request log entry.
Luckily, I've been able to reproduce the bug and find out the reason quite quickly.

Nonetheless, missing log entry was very unexpected.

@snoyberg how do you think real world apps should go about internal error logs? Do you suggest that it should keep track of all of its possible sources of exceptions and keep a separate log for that?

@mwotton do you need any help on this?

**Comment by mwotton** [comment]
I'm taking a few days off over Christmas - if you want to have a crack at
it feel free.

**Comment by snoyberg** [comment]
Applications should be catching all exceptions and converting them to 500
errors in a user friendly way (or machine friendly for JSON endpoints).
Blaming the logging middleware for not dealing with an application bug is
just ignoring the real problem.

Yesod installs these kinds of error handlers automatically, I'd recommend
that other web frameworks do the same.

**Comment by fizruk** [comment]
Ok, seems like the problem has moved to servant.

To reduce any confusion in the future I think the documentation for wai logger can be improved though.

**Comment by snoyberg** [comment]
A PR for improved docs would certainly be welcome.

**Comment by mwotton** [comment]
@snoyberg are haddock docs for Wai.Middleware.RequestLogger the appropriate place?

**Comment by snoyberg** [comment]
I don't have much of an opinion. I don't know what needs to be clarified
here exactly, so I'm not sure where such changes should go.

**Comment by saurabhnanda** [comment]
> Yesod installs these kinds of error handlers automatically, I'd recommend
that other web frameworks do the same.

Is there a reasonable use-case where someone would be using the WAI logger, but would NOT want this behaviour? Phrased another way, if every web framework using WAI logger would need to implement this behaviour, probably it makes sense for it to be WAI logger's default behaviour.

**Comment by Vlix** [comment]
#857 can probably help in this situation, right?

If you have the prelogged request, but not the actual accompanying request, that means there was probably an exception preventing the response.
